WAR-1813: Selenium: _make_browser() function got multiple values for keyword argument 'profile_dir'

safarheen opened the pull request:
Issue:
-------
_makebrowser method has profile_dir as well as **kw_args. However, profile_dir value is obtained from kw_args in previous step. Due to this, now there are two profile_dir attributes in __make_browser function call. This results in above failure.
TypeError: _make_browser() got multiple values for keyword argument 'profile_dir'

**Fix:**  Added logic to remove the profile_dir from kwargs, if exists.

Other Issues fixed along with this are:
Issue-1: 
-------
Binary, gecko_path, gecko_log, proxy_ip, proxy_port values given through datafile where not being read in browser_actions.

**Fix:**  Appended the default argument dict with these attributes. Now, the values are being read from the datafile.

Issue-2:
--------
gecko-log specified in datafile was not being read correctly in browser_management file. 

**Fix**: Updated the file to log the driver info in the path specified in datafile.

Issue-3:
--------
If empty/invalid binary path is passed in the datafile,  inappropriate error message is thrown.

**Fix:**  Provided appropriate error message if empty/invalid binary path is passed in the datafile.

Issue-4:
--------
ff_profile does not get set, if proxy_ip/port are None.

**Fix:**  Added else block to set ff_profile as profile_dir, in case proxy_ip/port are None.

Issue-5:
--------
get_firefox_version() keyword has a version check command specific to Linux only. Hence, execution in windows environment fails.

**Fix:** Added platform specific check in  get_firefox_version() keyword, such that, it builds the firefox version command based on the platform used.

travis.yml file update:
-----------------------
For Selenium 3, it makes sense to add tests with latest selenium version, latest geckodriver and a latest stable firefox version.
Hence,  updated the file.

Please find logs attached in the ticket: WAR-1813

BlackDuckCoPilot commented:
## Black Duck Security Report

Merging #425 into develop will not change security risk.

[Click here to see full report](https://copilot.blackducksoftware.com/github/repos/warriorframework/warriorframework/pull-requests/425)

---

1 component(s) did not change but have updated security data. This may affect your calculated score.

venkat1077 commented on warrior/Framework/ClassUtils/WSelenium/browser_mgmt.py:
```diff
@@ -538,6 +540,11 @@ def _make_ff(self, webdriver_remote_url, desired_capabilites, profile_dir, **kwa
                 # And firefox will need to know how to handle it
                 ff_capabilities['acceptInsecureCerts'] = True
+                
+                if not fileExists(str(binary)):
```
We need to handle two things here:
1. When user doesn't give any value for binary tag in IDF - Info_msg: No value given for binary, default firefox will be used
2. When the user provides some path but that doesn't exist - Warning_msg: Given binary path doesn't exist, default firefox will be used.

Pt. 2 is already there, just add Pt. 1.

safarheen replied:
Addressed the comments.

sanikakulkarni commented on warrior/Framework/ClassUtils/WSelenium/browser_mgmt.py:
```diff
@@ -515,15 +516,20 @@ def _make_ff(self, webdriver_remote_url, desired_capabilites, profile_dir, **kwa
+            ff_profile = profile_dir
         log_dir = get_object_from_datarepository("wt_logsdir") if \
-                  kwargs.get("log_path") in [None, False] else kwargs.get("log_path")
+                  gecko_log in [None, False] else gecko_log
```
Hi @safarheen,
Can you please explain the changes on line 532?

Thanks,
Sanika

sanikakulkarni replied:
Got it. Thanks!

sanikakulkarni commented:
Hi @safarheen,

Can you please explain the flow of the code change - that will help me a lot in doing the review right.

Thanks,
Sanika

safarheen commented on warrior/Actions/SeleniumActions/browser_actions.py:
```diff
@@ -230,13 +230,16 @@ def browser_launch(self, system_name, browser_name="all", type="firefox",
+            browser_optional_arg_keys = {"binary": None, "gecko_path": None, "proxy_ip": None, 
+                                         "proxy_port": None, "gecko_log": None}
+            # Adding browser_optional_arg_keys to arguments to get corresponding values from datafile.
+            arguments.update(browser_optional_arg_keys)
```
Hi @sanikakulkarni,

 binary, gecko_path, proxy_ip, proxy_port, and gecko_log are optional arguments, which will not be part of argument list initially, the line 232 with update the details of default arguments only(locals).

the values of  binary, gecko_path, proxy_ip, proxy_port, and gecko_log will be taken from the datafile in line 237 and updated into argument list.

sanikakulkarni replied:
You are right. This seems correct. Thanks!

safarheen commented:
Hi @sanikakulkarni,

These code changes ivolve fix for 4 issues.
The flow is as follows:
browser_actions file:
----------------------
Binary, gecko_path, gecko_log, proxy_ip, proxy_port values are considered as optional arguments which were not being read from datafile, as arguments is initialized to hold locals() which would be the default arguments. Hence, made a dict for opt args and updated these to argument list, so that values are fetched from datafile if any.

browser_mgmt.py:
--------------------
1. As kwargs was introduced, the profile_dir value was being passed twice, to fix this, added a changes to remove the profile_dir if present in kwagrs and pass the profile_dir value as such.
2. Firefox_proxy had an issue with setting preferences if proxy_ip and proxy_port are not passed. In order to give ability to set other preferences using profile_dir, added else block.
3. Initially the webdriver did not have a support to pass the logs of geckodriver. So, a default path was used (which is the execution dir).
Recently, the selenium webdriver supports to pass the log path where the geckodriver logs can be stored. (custamizable by the user) which can be passed through the datafile.
gecko_log attribute was added in browser_action file to implement this update. But, it was not replaced with the default log_path in browser_mngmnt file.
With this change, the gecko_log (path) if passed from datafile, will be considered. If not, the default execution dir will be used to save the geckodriver log.
4. To handle proper error message, if the firefox binary is passed empty. added a block to check if the passed binary value is appropriate, if not it would fetch the default firefox binary.

travis.yml:
----------
Added a test to validate that browser versions >47 work with geckodrivers installed.
Also, selenium V3 is used.

safarheen commented on warrior/Framework/ClassUtils/WSelenium/browser_mgmt.py:
```diff
@@ -539,6 +545,15 @@ def _make_ff(self, webdriver_remote_url, desired_capabilites, profile_dir, **kwa
+
                 # Force disable marionette, only needs in Selenium 3 with FF ver < 47
                 # Without these lines, selenium may encounter capability not found issue
                 # https://github.com/seleniumhq/selenium/issues/2739
```
Hi @sanikakulkarni 

Query:
-------
If we do not support Windows, it would affect the testing of Nagammai's team testing activities as they have been using both Linux and Windows for their test automation.
As per my knowledge, we do not have an OS independent command to get the firefox version.

Suggestion:
-------------
we could provide support, just for the version command update.
the following code would address this issue:

def get_firefox_version(self, binary):
        """
            Use firefox binary to find out firefox version
            before launching firefox in selenium
        """
        import platform
        raw_version=""
                   
        if platform.system() in "Linux":
            if binary in [False, None]:
                binary = "firefox"
            raw_version = check_output([binary, "-v"]) 
        elif platform.system() in "Windows":
            if binary in [False, None]:
                binary= FirefoxBinary._default_windows_location()
            print_info("*** ff binary for windows is : {} ".format(binary))
            command = "%s -v | more" % (binary)    
            raw_version = check_output(command)          
                
        version = False
        try:
            #raw_version = check_output(ff_version)
            print_info("*** Raw_version is: {} ".format(raw_version))
            match = re.search(r"\d+\.\d+", raw_version)
            if match is not None:
                version = LooseVersion(match.group(0))
            else:
                print_info("Cannot parse Firefox version: {}".format(raw_version))
        except CalledProcessError:
            print_error("Cannot find firefox version, will not launch browser")
        return version

sanikakulkarni replied:
Hi @safarheen,

This code looks fine. Please add it to the codebase (py2 and py3. Please note that the code *may* require changes when being ported to py3) and rerun the selenium projects and attach them to the JIRA ticket. Thanks.

safarheen replied:
Hi @sanikakulkarni 

I have added the platform check in the code for both py and py3. attached the tests in JIRA.
Please approve and merge the PRs

sanikakulkarni commented:
Selenium regression passed. Approving.
